annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java

```diff
@@ -0,0 +1,211 @@
+
+  // Needed for obtaining instances of utility classes Types and Elements.
+  // TODO: make methods non-static and pass source to constructor instead.
+  public static void setSource(Source source) {
+    if (InheritedSymbolFinder.source != source) {
+      InheritedSymbolFinder.source = source;
+      cache.clear();
+      cash.clear();
+      if (source != null) {
+        // initialize objectMethods
+        if (objectMethods == null) {
+          TypeElement o =
+              source.getElementUtils().getTypeElement("java.lang.Object");
+          objectMethods =
+              LinkedHashMultimap.<String, ExecutableElement>create();
+          for (Element e : o.getEnclosedElements()) {
+            if (e.getKind() == ElementKind.METHOD) {
+              String key = keyFor((Symbol.MethodSymbol) e);
+              objectMethods.put(key, (ExecutableElement) e);
+            }
+          }
+        }
+        cash.put("java.lang.Object", objectMethods);
+      }
+    }
+  }
+
+  /**
+   * Finds whether a method matching a JVML descriptor has an inherited
+   *  implementation in a given class.
+   *
+   * @param classSymbol symbol table entry for a class
+   * @param methodDescriptor JVML descriptor for a method
+   * @return whether a matching method exists in the class
+   */
+  public static boolean isInheritedIn(Symbol.ClassSymbol classSymbol,
+      String methodDescriptor) {
+    Multimap<String, ExecutableElement> inherited = getInherited(classSymbol);
+    int ix = methodDescriptor.indexOf('(');
+    String key = methodDescriptor.substring(0, ix);
+    Collection<ExecutableElement> elems = inherited.get(key);
+
+    // There may be multiple methods stored under the same key (bare
+    // method name), so check each of them.
+    // TODO: figure out good way to determine whether method descriptor
+    // gives the type of an inherited method.  (Only exact matches work
+    // now.)
+    if (elems != null) {
+      for (ExecutableElement elem : elems) {
+        String jvmName = JVMNames.getJVMMethodName(elem);
+        int ix0 = methodDescriptor.lastIndexOf(')') + 1;
+        int ix1 = jvmName.lastIndexOf(')') + 1;
+        String key0 = methodDescriptor.substring(0, ix0);
+        String key1 = jvmName.substring(0, ix1);
+        if (key0.equals(key1)) { return true; }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Finds inherited methods defined for a class (given by its symbol
+   *  table entry).
+   *
+   * @param classSymbol symbol table entry for a class
+   * @return JVML descriptors of methods in scope of class
+   */
+  private static Multimap<String, ExecutableElement>
+  getInherited(final Symbol.ClassSymbol classSymbol) {
+    if (source == null) { throw new IllegalStateException("null source"); }
+
+    final JavacElements elements = (JavacElements) source.getElementUtils();
+    final Types types = source.getTypeUtils();
+    String className = classSymbol.flatName().toString();
```

Comment:
Is there a reason that the previous two variables were final but this one is not?  The contrast makes a reader think there is a distinction being made.